src/infinitempomatrix.jl

```diff
@@ -44,111 +44,212 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
+    [dag(only(filterinds(A[x][1, 1]; plev=0, tags="Site"))) for x in 1:nsites(A)],
+    translator(A),
+  )
+end
+
+function ITensors.splitblocks(H::InfiniteMPOMatrix)
+  N = nsites(H)
+  for j in 1:N
+    for n in 1:length(H)
+      H[j][n] = splitblocks(H[j][n])
+    end
+  end
+  return H
+end
+
+function find_all_links(Hm::Matrix{ITensor})
+  is = inds(Hm[1, 1]) #site inds
   lx, ly = size(Hm)
-  @assert lx == ly
-  #
-  #  Extract the sub diagonal
-  #
-  Ms = map(n -> Hm[lx - n + 1, ly - n], 1:(lx - 1)) #Get the sub diagonal into an array.
-  N = length(Ms)
-  @assert N == lx - 1
-  #
-  # Convert edge Ms to order 4 ITensors using the dummy index.
-  #
-  il0 = inds_array[1][1]
-  T = eltype(Ms[1])
-  Ms[1] *= onehot(T, il0 => 1)
-  Ms[N] *= onehot(T, dag(il0) => 1) #We don't need distinct index IDs for this.  directsum makes all new index anyway.
-  #
-  # Start direct sums to build the single ITensor.  Order is critical to get the desired form.
-  #
-  H = Hm[1, 1] * onehot(T, il0' => 1) * onehot(T, dag(il0) => 1) #Bootstrap with the I op in the top left of Hm
-  H, il = directsum(H => il0', Ms[N] => inds_array[N][1]) #1-D directsum to put M[N] directly below the I op
-  ir = dag(il0) #setup recusion.
-  for i in (N - 1):-1:1 #2-D directsums to place new blocks below and to the right.
-    H, (il, ir) = directsum(
-      H => (il, ir), Ms[i] => inds_array[i]; tags=["Link,left", "Link,right"]
+  #We extract the links from the order-3 tensors on the first column and line
+  #We add dummy indices if there is no relevant indices
+  ir = only(uniqueinds(Hm[1, 2], is))
+  ir0 = Index(ITensors.trivial_space(ir); dir=dir(ir), tags="Link,extra")
+  il0 = dag(ir0)
+  left_links = typeof(ir)[]
+  for x in 1:lx
+    temp = uniqueinds(Hm[x, 1], is)
+    if length(temp) == 0
+      append!(left_links, [il0])
+    elseif length(temp) == 1
+      append!(left_links, temp)
+    else
+      error("ITensor does not seem to be of the correct order")
+    end
+  end
+  right_links = typeof(ir)[]
+  for x in 1:lx
+    temp = uniqueinds(Hm[1, x], is)
+    if length(temp) == 0
+      append!(right_links, [ir0])
+    elseif length(temp) == 1
+      append!(right_links, temp)
+    else
+      error("ITensor does not seem to be of the correct order")
+    end
+  end
+  return left_links, right_links
+end
+
+function convert_itensor_to_itensormatrix(tensor; kwargs...)
+  if order(tensor) == 3
+    return convert_itensor_3vector(tensor; kwargs...)
+  elseif order(tensor) == 4
+    return convert_itensor_33matrix(tensor; kwargs...)
+  else
+    error(
+      "Conversion of ITensor into matrix of ITensor not planned for this type of tensors"
     )
   end
-  IN = Hm[N + 1, N + 1] * onehot(T, dag(il0) => 1) * onehot(T, il => dim(il)) #I op in the bottom left of Hm
-  H, ir = directsum(H => ir, IN => il0; tags="Link,right") #1-D directsum to the put I in the bottom row, to the right of M[1]
+end
+
+"""
+    build_three_projectors_from_index(is::Index; kwargs...)
```

Comment:
What about `local_mpo_block_projectors`?

Reply:
To indicate what the projectors are being used for, i.e. to transform an MPO tensor into a local MPO block structure.